main/src/main/scala/sbt/Main.scala

```diff
@@ -13,25 +13,56 @@ import scala.annotation.tailrec
 import Path._
 import StandardMain._
 
-import java.io.File
+import java.io.{ File, IOException }
 import java.net.URI
 import java.util.Locale
 import scala.util.control.NonFatal
 
-/** This class is the entry point for sbt.*/
+/** This class is the entry point for sbt. */
 final class xMain extends xsbti.AppMain {
   def run(configuration: xsbti.AppConfiguration): xsbti.MainResult =
     {
       import BasicCommands.early
       import BasicCommandStrings.runEarly
       import BuiltinCommands.{ initialize, defaults }
       import CommandStrings.{ BootCommand, DefaultsCommand, InitCommand }
+      if (!java.lang.Boolean.getBoolean("sbt.skip.version.write"))
+        setSbtVersion(configuration.baseDirectory(), configuration.provider().id().version())
       runManaged(initialState(configuration,
         Seq(defaults, early),
         runEarly(DefaultsCommand) :: runEarly(InitCommand) :: BootCommand :: Nil)
       )
     }
+
+  private val sbtVersionRegex = """sbt\.version\s*=.*""".r
+  private def isSbtVersionLine(s: String) = sbtVersionRegex.pattern matcher s matches ()
+
+  private def isSbtProject(baseDir: File, projectDir: File) =
+    projectDir.exists() || (baseDir * "*.sbt").get.nonEmpty
+
+  private def setSbtVersion(baseDir: File, sbtVersion: String) = {
+    val projectDir = baseDir / "project"
+    val buildProps = projectDir / "build.properties"
+
+    val buildPropsLines = if (buildProps.canRead) IO.readLines(buildProps) else Nil
+
+    val sbtVersionAbsent = buildPropsLines forall (!isSbtVersionLine(_))
+
+    if (sbtVersionAbsent) {
+      val errorMessage = s"WARN: No sbt.version set in project/build.properties, base directory: $baseDir"
+      try {
+        if (isSbtProject(baseDir, projectDir)) {
+          val newBuildPropsLines = s"sbt.version=$sbtVersion" :: buildPropsLines
+          IO.writeLines(buildProps, newBuildPropsLines)
```

Comment:
I think sbt should tell the user that it's writing the file. Software should not perform side effects silently without explicit user consent.